### PR TITLE
feat: add animation prop to bottom tab

### PR DESCRIPTION
**Pull request by teneeto**
**Description**
----

This PR adds  properties to `bottom-tabs` navigation `screenOptions` to enable animation (as `animationEnabled`),  add custom animations or use presets (as `styleInterpolator`), and also let you set  transition config (as `transitionSpec`) hence, providing a medium to animate bottoms-tabs screens.

**Video Walkthrough**
----
The tabs as shown in the video currently uses a custom `fade` preset.

https://user-images.githubusercontent.com/22779249/234637219-9456967d-55c7-4318-9c7e-41a842747c12.mov

**Quality check**
----
The change takes into account  lint, typescript and test checks (all passing).

### CODE REVIEW

File: example/src/Screens/BottomTabsAnimated.tsx [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          tabBarBackground: () => (
+            <BlurView
+              tint="dark"
+              intensity={100}
+              style={StyleSheet.absoluteFill}
+            />
+          ),
+        }}
+      />
+    </Tab.Navigator>
```
Comment (satya164) [inline example/src/Screens/BottomTabsAnimated.tsx:140]: This navigator can be simplified by removing any customization unrelated to animations (such as blurred tab bar).

**Review by satya164 — CHANGES_REQUESTED** [review]
Currently the cross-fade animation is broken. It's more obvious if you slow down the animation. Here I have slowed down the animation by 10x:

https://user-images.githubusercontent.com/1174278/233005257-7b78d7ac-dabd-4af0-9880-4f47d0215aed.mp4

With a cross-fade animation, both of the screens need to stay visible. But currently the unfocused screen always disappears and only the focused screen fades in.

Here's how it should behave:


https://user-images.githubusercontent.com/1174278/233318845-36ca9e1f-bd8b-4b48-ae41-04aefc4d7ac6.mp4



There are 2 reasons for this:

- Inactive screens get detached using `MaybeScreen` and `MaybeScreenContainer`. They need to be kept around for the duration of the animation. This can be achieved by `hasTwoStates={false}` and setting `visible={true}` for the 2 screens which are currently animating.
- The `Screen` component renders `Background` which will render a opaque background hiding the screen underneath. You need to pass `backgroundColor: 'transparent'` and handled the background color to be `colors.background` in the inner view that animates.

File: packages/bottom-tabs/src/views/BottomTabView.tsx [hunk 2]
```diff
@@ -45,12 +46,45 @@ export function BottomTabView(props: Props) {
...
+  const tabsPositionAnims = useAnimatedValueObject(state.routes);
+
+  const animateToIndex = React.useCallback(() => {
+    state.routes.map((route) => {
+      return Animated[transitionSpec?.animation || 'timing'](
+        tabsPositionAnims[route.key],
+        {
+          toValue: route.key === focusedRouteKey ? 0 : 1,
+          useNativeDriver: true,
+          ...transitionSpec?.config,
```
[the remark this answers is not in the document]
Comment (teneeto) [reply packages/bottom-tabs/src/views/BottomTabView.tsx:72]: Although they are already omitted, i will implement them for safety.

[the remark this answers is not in the document]
Comment (satya164) [reply packages/bottom-tabs/src/views/BottomTabView.tsx:72]: They are only omitted in the types. But it won't prevent users from passing them, especially if they don't use TypeScript.

**Comment by andreialecu** [comment]
This looks great, and I'm very interested in this feature. 

@teneeto any plans to finish it?

**Comment by teneeto** [comment]
> This looks great, and I'm very interested in this feature.
> 
> @teneeto any plans to finish it?

Yes, this is in Review. Your contributions would mean a lot too.

**Comment by osdnk** [comment]
Hi @teneeto, thanks for your contribution and addressing our review. 
Before we can proceed, I need to share one more concern I found out while testing your changes, 
Before: 

https://github.com/react-navigation/react-navigation/assets/25709300/d75de053-cc4c-46d3-9a98-10b9a5f2a6e7
After: 

https://github.com/react-navigation/react-navigation/assets/25709300/a7ca396c-1197-4d12-b7be-4d0368d3aa44

It looks like something is definitely off. That happens only with "regular" bottom tabs. 

Also, we're trying to keep one example per feature, which exhaustively presents all capabilities. Is there a chance we can merge that example into regular bottom-tabs? If yes, we can add a which to turn on/off the preset.

**Review by osdnk — CHANGES_REQUESTED** [review]
Unfortunately, this is blocking to my understanding.

**Comment by osdnk** [comment]
@teneeto I think i investigated bugs. Adjusting the example is the remaining part

**Comment by teneeto** [comment]
> osdnk

Hi Michał, thanks for finding some time to go through this, I'll look into what needs to be improved as soon as I can.

**Comment by osdnk** [comment]
I only encourage you not to create a new example on the main screen but to integrate it with an existing bottom-tabs example (with a button to turn the animation on and off).

**Review by osdnk — APPROVED** [review]

**Comment by osdnk** [comment]
I think I'm happy with those changes. I had some issues with flickering with `hasTwoStates` as screens' container. I did a lazy thing and disabled that if any screen has `animationEnabled` (otherwise I faced extra-rerender). It's not changing anything if the we don't use animation.

**Pull request merged**